presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -126,6 +129,41 @@ public static BindableAggregationFunction generateAggregationBindableFunction(Cl
         return builder.build();
     }
 
+    private static List<TypeVariableConstraint> parseTypeVariableConstraints(Method inputFunction)
+    {
+        return createTypeVariableConstraints(Arrays.asList(inputFunction.getAnnotationsByType(TypeParameter.class)), parseImplementationDependencies(inputFunction));
+    }
+
+    private static List<ScalarImplementation.ImplementationDependency> parseImplementationDependencies(Method inputFunction)
+    {
+        ImmutableList.Builder<ScalarImplementation.ImplementationDependency> builder = ImmutableList.builder();
+        List<TypeParameter> typeParameters = Arrays.asList(inputFunction.getAnnotationsByType(TypeParameter.class));
+        Set<String> literalParameters = getLiteralParameter(inputFunction);
+
+        for (int i = 0; i < inputFunction.getParameterCount(); i++) {
+            Class<?> parameterType = inputFunction.getParameterTypes()[i];
+            Annotation[] annotations = inputFunction.getParameterAnnotations()[i];
+
+            // Skip injected parameters
+            if (parameterType == ConnectorSession.class) {
+                continue;
+            }
+
+            if (containsMetaParameter(annotations)) {
+                checkArgument(annotations.length == 1, "Meta parameters may only have a single annotation [%s]", inputFunction);
+                Annotation annotation = getOnlyElement(Arrays.asList(annotations));
+                if (annotation instanceof TypeParameter) {
+                    checkArgument(typeParameters.contains(annotation), "Injected type parameters must be declared with @TypeParameter annotation on the method [%s]", inputFunction);
```

Comment:
We changed this check on the scalar side recently. We now allow one to inject types like `map(varchar, bigint)` or `map(varchar, T)`.